src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java

```diff
@@ -338,42 +363,24 @@ void dropIndexSSTables(Collection<SSTableReader> ss, StorageAttachedIndex index)
     public synchronized Set<StorageAttachedIndex> onSSTableChanged(Collection<SSTableReader> removed, Iterable<SSTableReader> added,
                                                             Set<StorageAttachedIndex> indexes, boolean validate)
     {
-        Pair<Set<SSTableContext>, Set<SSTableReader>> results = contextManager.update(removed, added, validate);
+        Optional<Set<SSTableContext>> optValid = contextManager.update(removed, added, validate);
 
-        if (!results.right.isEmpty())
+        if (optValid.isEmpty())
         {
-            results.right.forEach(sstable -> {
-                IndexDescriptor indexDescriptor = IndexDescriptor.createFrom(sstable);
-                indexDescriptor.deletePerSSTableIndexComponents();
```

Comment:
It's been moved. It's now directly in `IndexDescriptor.ComponentGroupImpl#validateComponents`: if we detect corruption there, we delete the corrupted files as appropriate.

Which I did for a couple reasons:
1. it felt cleaner to me to have the handling of corruption as close from the detection of corruption as possible. Makes it easier to have slightly more helpful messages for ones (with details on what is corrupted).
2. it happens to avoid code duplication between "per-sstable" and "per-index" stuffs. It would also save duplication if we start calling validation elsewhere.
3. I handle it slightly differently with new immutable components than in "legacy" mode. In legacy, we delete the whole "group" as soon as any component is corrupted, but with immutable, I delete only the corrupted stuff and the completion marker (the idea being, don't delete anything that may be actively read unless it's provably broken). Having that difference better encapsulated felt cleaner too.